### ISSUE TITLE
Pre-trained model for Efficient net B0 implementation

### ISSUE THREAD

**Issue by peri044** [tensorflow/models#9054]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [X] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [X] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [X] I checked to make sure that this issue has not been filed already.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/official/vision/image_classification

## 2. Describe the bug

For efficient net implementation (classifier_trainer.py), i the [TFHUB](https://tfhub.dev/tensorflow/efficientnet/b0/classification/1 ) saved model or pretrained weights expected to work ? I'm a bit confused as the [README.md](https://github.com/tensorflow/models/tree/master/official/vision/image_classification#efficientnet) says efficient net is work in progress.

I'm evaluating the saved_model.pb provided in the tfhub for B0 model and my accuracy on Imagenet is close to zero. 

The way I'm evaluating is as follows:
In the classifer_trainer.py, I add `model = tf.keras.models.load_model('eff_b0_checkpoint/')` (directory which has saved_model.pb)
and comment out the training part and skip to 

> validation_output = model.evaluate(validation_dataset, steps=validation_steps, verbose=2)

Can you please let me know if this is right ? Thank you !!

**Comment by allenwang28** [comment]
For clarification, EfficientNet is a WIP in that we have only been able to successfully train EfficientNet-b0 and b1, but haven't had cycles to scale up. 

The TFHUB pre-trained weights should work correctly. Can you ensure that [this codepath](https://github.com/tensorflow/models/blob/r2.3.0/official/vision/image_classification/classifier_trainer.py#L193) is not run by appending `--params_override="train.resume_checkpoint=False"` to your command?

What is the validation accuracy you're seeing?

**Comment by peri044** [comment]
Thank you for looking into this. Yes. That codepath is commented out. 
I'm directly loading the saved model first and evaluating directly before model.fit in the classifer_trainer.py

> model = tf.keras.models.load_model('eff_b0_checkpoint/')
> model.compile(optimizer .....)
> validation_output = model.evaluate(validation_dataset, steps=validation_steps, verbose=2)
> print (validation_output)
loss: 7000.2163 - accuracy: 0.0012 - top_5_accuracy: 0.9857

Command used for eval:
`python3 classifier_trainer.py --mode=train_and_eval --model_type=efficientnet --dataset=imagenet --model_dir=expt1 --data_dir=/mnt/cdr/ImageNet/train-val-tfrecord/ --config_file=configs/examples/efficientnet/imagenet/efficientnet-b0-gpu.yaml`
In the efficientnet config, I just changed `distribution_strategy` to be `off`. I'm using the default autoaugment option. Is there any specific preprocessing related info that need to be changed?

**Comment by allenwang28** [comment]
Hmm, I think the preprocessing *should* be fine by default since we mean subtract and standardize. I think it's possible that the TFHub model doesn't use one hot encoding but we do. Can you try changing [this](https://github.com/tensorflow/models/blob/r2.3.0/official/vision/image_classification/configs/examples/efficientnet/imagenet/efficientnet-b0-gpu.yaml#L47) to 0. instead and see if this helps?

**Comment by peri044** [comment]
Changed it but the issue still exists. 
Here is the reference code if it helps
[https://gist.github.com/peri044/e0278249f8a8a13e47c0a9a9e58a007f#file-ctl_2-py-L391](https://gist.github.com/peri044/e0278249f8a8a13e47c0a9a9e58a007f#file-ctl_2-py-L391)

Command:
`python3 classifier_trainer.py --mode=train_and_eval --model_type=efficientnet --dataset=imagenet --model_dir=expt1 --data_dir=/mnt/cdr/ImageNet/train-val-tfrecord/ --config_file=configs/examples/efficientnet/imagenet/efficientnet-b0-gpu.yaml`

**Comment by allenwang28** [comment]
Thanks for the reference code! I just remembered that our EfficientNet model has an option to rescale inputs in the model so we don't enable this by default.

Can you enable `mean_subtract` and `standardize` on `train_dataset` and `validation_dataset` within [efficientnet-b0-gpu.yaml](https://github.com/tensorflow/models/blob/r2.3.0/official/vision/image_classification/configs/examples/efficientnet/imagenet/efficientnet-b0-gpu.yaml#L8)? You can refer to [the ResNet GPU](https://github.com/tensorflow/models/blob/r2.3.0/official/vision/image_classification/configs/examples/resnet/imagenet/gpu.yaml) as an example.

**Comment by peri044** [comment]
Thanks for the info. I did that and 
1) Loss changed a lot from  
loss: 7000.2163 - accuracy: 0.0012 - top_5_accuracy: 0.9857 (without mean_subtract and standardize )
   to
loss: 9.6760 - accuracy: 0.0042 - top_5_accuracy: 0.0133 ( with both of them)
Surprisingly the accuracy is still worse.
Just a note (may not be helpful): The loss value of 9.67 previously (for resnet ) indicated that the model is uninitialized network. But now since I'm loading efficient net saved model, this does seem weird. 
Any suggestions ? 

2) One more question. You've mentioned in the above comments that EfficientNet is WIP and only B0 and B1 are successfully trained. But there is a whole collection of models [here](https://tfhub.dev/google/collections/efficientnet/1) from B0 to B7. Does it mean they work with this codebase ? Thanks

**Comment by allenwang28** [comment]
1. Please try turning label smoothing back on by undoing my suggestion:

> Hmm, I think the preprocessing _should_ be fine by default since we mean subtract and standardize. I think it's possible that the TFHub model doesn't use one hot encoding but we do. Can you try changing [this](https://github.com/tensorflow/models/blob/r2.3.0/official/vision/image_classification/configs/examples/efficientnet/imagenet/efficientnet-b0-gpu.yaml#L47) to 0. instead and see if this helps?

I think there's possibly a few thing still misconfigured which is why we see such a high loss/low accuracy. You might try toggling variations of one hot encoding, mean subtract and standardization - we should see around 0.767-0.768 if we did it correctly.

2. The WIP means that we haven't trained from scratch from B2-B7 successfully. The TFHub models should work fine with this codebase though.

**Comment by peri044** [comment]
@allenwang28 
1) I was able to fix the issue. In the saved_model provided in TFhub, in addition to mean_subtract and standardize, there is a `mul` layer (as indicated in the screenshot which is weird) which multiplies the input by 255. So I disabled mean subtract, standardize (since they are already in the saved_model graph) and downscaled the input by 255.  Now I get 76.8 % accuracy.
![image](https://user-images.githubusercontent.com/10511428/89595914-bc617e80-d80a-11ea-9b2d-06c005b232dc.png)

2) So far, I used the provided saved model. But I tried loading the weights and in both the cases, the weights failed to load saying the objects didn't match.

Case 1: 
    
> model = get_models()[params.model.name](**model_params)
> model.load_weights('./eff_b0/variables/variables')

Case 2: 
> model = get_models()[params.model.name](**model_params)
> checkpoint = tf.train.Checkpoint(model=model, optimizer=optimizer)
> checkpoint.restore('./eff_b0/variables/variables').assert_consumed()

Any suggestions here ? Thank you !!

**Comment by allenwang28** [comment]
Great, glad you were able to figure it out!

It looks like the intended usage for TFHub is to use `hub.KerasLayer` ([example here](https://www.tensorflow.org/hub/tutorials/tf2_image_retraining#defining_the_model)) instead of loading weights like a checkpoint.

Will this work for your use case?

**Comment by peri044** [comment]
@allenwang28 Sorry for the delay. Had to switch tasks. I figured how to use it for fine-tuning. Thanks for the info.
